### PR TITLE
The rest patches for support IMX8ULP

### PR DESCRIPTION
**Pull request by TangleZ**

### CODE REVIEW

File: src/include/sof/platform.h [hunk 1]
```diff
@@ -54,6 +54,24 @@ int platform_context_save(struct sof *sof);
...
+	case HOST_TO_LOCAL:
+		*addr = SDRAM0_BASE + (*addr - MEM_RESERVED);
+		break;
+	case LOCAL_TO_HOST:
+		*addr = MEM_RESERVED + (*addr - SDRAM0_BASE);
+		break;
+	};
+#endif
+}
+#else
```
Comment (lyakh) [inline src/include/sof/platform.h:73]: this `#else` doesn't seem to be needed

File: src/include/sof/platform.h [hunk 2]
```diff
@@ -54,6 +54,24 @@ int platform_context_save(struct sof *sof);
...
+	switch (type) {
+	case HOST_TO_LOCAL:
+		*addr = SDRAM0_BASE + (*addr - MEM_RESERVED);
+		break;
+	case LOCAL_TO_HOST:
+		*addr = MEM_RESERVED + (*addr - SDRAM0_BASE);
+		break;
+	};
+#endif
+}
```
Comment (lyakh) [inline src/include/sof/platform.h:72]: seems like too many conditions to me. Functions like
```
void fn(bool mode)
{
	if (mode)
		do_one_thing();
	else
		do_another_thing();
}
```
which are then mostly called as `fn(true)` or `fn(false)` seem counterproductive to me - just call the `do_x_thing()` directly! Which here would mean - I'd define two distinct functions or macros `convert_local_to_host()` and `convert_host_to_local()` and just call them directly

**Review by lyakh — CHANGES_REQUESTED** [review]

File: src/drivers/imx/sai.c [hunk 3]
```diff
@@ -31,6 +31,7 @@ static void sai_start(struct dai *dai, int direction)
 {
 	dai_info(dai, "SAI: sai_start");
```
Comment (dbaluta) [inline src/drivers/imx/sai.c:32]: There are 2 different changes in this patch:
1. support for 16bit mono which results in changes for sywd and twd
2. support for SAI lanes selection (tx0, rx2)

Please split this patch into 2 and provide a better explanation.

File: src/include/sof/drivers/sai.h [hunk 4]
```diff
@@ -18,6 +18,8 @@
 
 #ifdef CONFIG_IMX8M
 #define SAI_OFS		8
```
Comment (dbaluta) [inline src/include/sof/drivers/sai.h:20]: can you add here
```
#if defined(CONFIG_IMX8M) ||  defined(CONFIG_IMX8ULP)
#define   SAI_OF 8
#else    
#define SAI_OFS 0
```

https://stackoverflow.com/questions/9682593/how-to-use-ifdef-with-an-or-condition

File: tools/topology/topology1/CMakeLists.txt [hunk 5]
```diff
@@ -138,6 +138,7 @@ set(TPLGS
 	"sof-imx8-cs42888-mixer\;sof-imx8-cs42888-mixer"
 	"sof-imx8mp-wm8960-mixer\;sof-imx8mp-wm8960-mixer"
 	"sof-imx8mp-wm8960-kwd\;sof-imx8mp-wm8960-kwd"
+	"sof-imx8ulp-btsco\;sof-imx8ulp-btsco"
```
Comment (dbaluta) [inline tools/topology/topology1/CMakeLists.txt:141]: This should be kept the last patch in the series.

File: tools/topology/topology1/sof/pipe-volume-playback.m4 [hunk 6]
```diff
@@ -80,7 +80,8 @@ indir(`define', concat(`PIPELINE_PCM_', PIPELINE_ID), Passthrough Playback PCM_I
 # PCM Configuration
 
 #
-PCM_CAPABILITIES(Passthrough Playback PCM_ID, CAPABILITY_FORMAT_NAME(PIPELINE_FORMAT), PCM_MIN_RATE, PCM_MAX_RATE, 2, PIPELINE_CHANNELS, 2, 16, 192, 16384, 65536, 65536)
+PCM_CAPABILITIES(Passthrough Playback PCM_ID, CAPABILITY_FORMAT_NAME(PIPELINE_FORMAT),
+			PCM_MIN_RATE, PCM_MAX_RATE, PIPELINE_CHANNELS, PIPELINE_CHANNELS, 2, 16, 192, 16384, 65536, 65536)
```
Comment (dbaluta) [inline tools/topology/topology1/sof/pipe-volume-playback.m4:85]: Is this OK, you seem to add an extra parameter to PCM_CAPABILITIES?

Comment (TangleZ) [reply tools/topology/topology1/sof/pipe-volume-playback.m4:85]: No, just instead 2 by PIPELINE_CHANNELS, it's means channel_min

File: src/include/sof/platform.h [hunk 7]
```diff
@@ -54,6 +54,24 @@ int platform_context_save(struct sof *sof);
 
 /** @}*/
```
Comment (dbaluta) [inline src/include/sof/platform.h:56]: @lgirdwood @plbossart @lyakh can you please review as it touches the core. 

Our hardware (i.MX8ULP) has an interesting mapping for some shared memory.

DSP physical address range:

```0x1a000000~0x1bffffff```

is mapped to ARM physical address range:
```
0x8e000000~0x8fffffff 
```
so we need somehow to do a `translation` of the addresses.

File: src/include/sof/platform.h (hunk 1 above)
Comment (TangleZ) [reply src/include/sof/platform.h:73]: Sure. Will remove this.

File: tools/topology/topology1/sof/pipe-volume-playback.m4 (hunk 6 above)
Comment (dbaluta) [reply tools/topology/topology1/sof/pipe-volume-playback.m4:85]: True. You are right.

File: tools/topology/topology1/sof/pipe-volume-playback.m4 [hunk 8]
```diff
@@ -80,7 +80,8 @@ indir(`define', concat(`PIPELINE_PCM_', PIPELINE_ID), Passthrough Playback PCM_I
 # PCM Configuration
```
Comment (dbaluta) [inline tools/topology/topology1/sof/pipe-volume-playback.m4:81]: Small checkpatch.pl warning on this patch:

```
WARNING: 'paramter' may be misspelled - perhaps 'parameter'?
```

File: src/include/sof/platform.h (hunk 1 above)
Comment (dbaluta) [reply src/include/sof/platform.h]: We left this out of this PR because it is sensible and it needs careful review from more people.

**Comment by dbaluta** [comment]
Looks good now. Please submit the patches that handle the conversion of addresses between Kernel and Firmware in a separate PR.

**Review by dbaluta — APPROVED** [review]

File: src/include/sof/drivers/sai.h [hunk 9]
```diff
@@ -237,6 +237,14 @@
 
 #define SAI_FLAG_PMQOS   BIT(0)
 
+#ifdef CONFIG_IMX8M
+#define SAI_FIFO_WORD_SIZE	128
+#elif defined CONFIG_IMX8ULP
```
Comment (paulstelian97) [inline src/include/sof/drivers/sai.h:242]: I'd have gone for #elifdef but this is also fine.

File: tools/topology/topology1/CMakeLists.txt [hunk 10]
```diff
@@ -138,6 +138,7 @@ set(TPLGS
 	"sof-imx8-cs42888-mixer\;sof-imx8-cs42888-mixer"
```
Comment (paulstelian97) [inline tools/topology/topology1/CMakeLists.txt:138]: I don't see in the last patch what "Use edma copy data." means. You can remove that line from the commit message.

**Review by paulstelian97 — APPROVED** [review]
Mostly good, just a few things to nitpick here.

**Comment by dbaluta** [comment]
@lyakh we left out the patches that touch SOF core. Can you have another look. We will post those patches separately as they try to address a sensible issue.

File: tools/topology/topology1/CMakeLists.txt (hunk 10 above)
Comment (TangleZ) [reply tools/topology/topology1/CMakeLists.txt:138]: Add this just for understanding how the pipeline works on 8ulp. If you think we don't need it, I will remove it.

Comment (paulstelian97) [reply tools/topology/topology1/CMakeLists.txt:138]: The DMA engine isn't that relevant when describing a pipeline, shouldn't be part of the specification of the pipeline itself. I'd say remove that line.

Comment (TangleZ) [reply tools/topology/topology1/CMakeLists.txt:138]: Removed.

**Review by paulstelian97 — APPROVED** [review]

File: tools/topology/topology1/sof/pipe-volume-playback.m4 [hunk 11]
```diff
@@ -80,7 +80,8 @@ indir(`define', concat(`PIPELINE_PCM_', PIPELINE_ID), Passthrough Playback PCM_I
...
 
 #
-PCM_CAPABILITIES(Passthrough Playback PCM_ID, CAPABILITY_FORMAT_NAME(PIPELINE_FORMAT), PCM_MIN_RATE, PCM_MAX_RATE, 2, PIPELINE_CHANNELS, 2, 16, 192, 16384, 65536, 65536)
+PCM_CAPABILITIES(Passthrough Playback PCM_ID, CAPABILITY_FORMAT_NAME(PIPELINE_FORMAT),
+			PCM_MIN_RATE, PCM_MAX_RATE, PIPELINE_CHANNELS, PIPELINE_CHANNELS, 2, 16, 192, 16384, 65536, 65536)
```
Comment (lgirdwood) [inline tools/topology/topology1/sof/pipe-volume-playback.m4:84]: This is to support playing stereo on a 4 channel pipeline.

**Review by lgirdwood — CHANGES_REQUESTED** [review]
If you need a fixed min=max chan topology then it may be better to pass this is as an argument to the topology generation.

File: tools/topology/topology1/sof/pipe-volume-playback.m4 (hunk 11 above)
Comment (TangleZ) [reply tools/topology/topology1/sof/pipe-volume-playback.m4:84]: If compared with pipe-volume-capture.m4, you can find it also fixed min = max, and this, for now, should be the same with hardware, so I think it should be min channel = max channel. And here is a definition, you can set channel is 4 in your topology file.

**Comment by dbaluta** [comment]
@lgirdwood actually we need min chan = 1, with the current code min chan is always forced to value 2. Let us try to find an elegant solution for this.

File: tools/topology/topology1/sof/pipe-volume-playback.m4 (hunk 11 above)
Comment (dbaluta) [reply tools/topology/topology1/sof/pipe-volume-playback.m4:84]: @TangleZ please remove this commit from the current PR, and will send a separate patch.

Comment (TangleZ) [reply tools/topology/topology1/sof/pipe-volume-playback.m4]: Removed.

Comment (TangleZ) [reply tools/topology/topology1/sof/pipe-volume-playback.m4]: @lgirdwood Removed this change, will find a better way to address this.

**Comment by dbaluta** [comment]
@lgirdwood @plbossart this is ready to be merged. We will greatly appreciate your feedback as soon as possible as we need this in to kick our build internal build process.

**Review by plbossart — APPROVED** [review]

**Review by lgirdwood — APPROVED** [review]

**Pull request merged**